packages/@uppy/companion-client/src/RequestClient.ts

```diff
@@ -0,0 +1,613 @@
+'use strict'
+
+import UserFacingApiError from '@uppy/utils/lib/UserFacingApiError'
+// eslint-disable-next-line import/no-extraneous-dependencies
+import pRetry, { AbortError } from 'p-retry'
+
+import fetchWithNetworkError from '@uppy/utils/lib/fetchWithNetworkError'
+import ErrorWithCause from '@uppy/utils/lib/ErrorWithCause'
+import emitSocketProgress from '@uppy/utils/lib/emitSocketProgress'
+import getSocketHost from '@uppy/utils/lib/getSocketHost'
+
+import type Uppy from '@uppy/core'
+import type { UppyFile, Meta, Body } from '@uppy/utils/lib/UppyFile'
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-ignore untyped because we're getting rid of it
+import type { RateLimitedQueue } from '@uppy/utils/lib/RateLimitedQueue'
+import AuthError from './AuthError.ts'
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-ignore We don't want TS to generate types for the package.json
+import packageJson from '../package.json'
+
+type CompanionHeaders = Record<string, string> | undefined
+
+export type Opts = {
+  name?: string
+  provider: string
+  pluginId: string
+  companionUrl: string
+  companionCookiesRule?: 'same-origin' | 'include' | 'omit'
+  companionHeaders?: CompanionHeaders
+  companionKeysParams?: Record<string, string>
+}
+
+export type RequestOptions =
+  | boolean
+  | {
+      method?: string
+      data?: Record<string, unknown>
+      skipPostResponse?: boolean
+      signal?: AbortSignal
+      qs?: Record<string, string>
+    }
+
+// Remove the trailing slash so we can always safely append /xyz.
+function stripSlash(url: string) {
+  return url.replace(/\/$/, '')
+}
+
+const retryCount = 10 // set to a low number, like 2 to test manual user retries
+const socketActivityTimeoutMs = 5 * 60 * 1000 // set to a low number like 10000 to test this
+
+export const authErrorStatusCode = 401
+
+class HttpError extends Error {
+  statusCode: number
+
+  constructor({
+    statusCode,
+    message,
+  }: {
+    statusCode: number
+    message: string
+  }) {
+    super(message)
+    this.name = 'HttpError'
+    this.statusCode = statusCode
+  }
+}
+
+async function handleJSONResponse<JSON extends Record<string, unknown>>(
+  res: Response,
+): Promise<JSON> {
```

Comment:
```suggestion
): Promise<ReturnType<JSON.parse>> {
```

Reply:
Oh JSON is a generic, that's confusing. Can we use a different name?
```suggestion
): Promise<JSONData> {
```

Reply:
I think it's okay, `handleJSONResponse` with a `JSON` generic, it seems quite clear what it corresponds too. My personal pet peeve is calling things `data`, a meaningless word in programming as everything is data.

Reply:
I can hear that "data" does not correspond to much, but on the other hand `JSON` usually refers to `globalThis.JSON`, so yes, please let's use a different name, I'm fine with whatever.